### PR TITLE
DCS-872 supporting DISPUTED statements 

### PR DESCRIPTION
**Pull request by andrewrlee**
and tidying up retrieval of reviewer statements

### CODE REVIEW

File: server/data/statementsClient.test.ts [hunk 1]
```diff
@@ -40,22 +40,22 @@ test('getStatementForUser', () => {
...
+            ,      s.name                   "name"
+            ,      r.reporter_name          "reporterName"
+            from report r
+            left join statement s on r.id = s.report_id
+            where r.id = $1
+              and r.deleted is null
+              and s.user_id = $2
+              and s.statement_status = $3
+              and s.deleted is null`,
     values: [19, 'user-1', StatementStatus.PENDING.value],
```
Comment (andrewrlee) [inline server/data/statementsClient.test.ts:59]: just white space changes

File: server/data/statementsClient.ts [hunk 2]
```diff
@@ -41,61 +41,74 @@ export default class StatementsClient {
...
+            from v_report r
+            left join v_statement s on r.id = s.report_id
+            where s.id = $3`,
+      values: [StatementStatus.SUBMITTED.value, StatementStatus.DISPUTED.value, statementId],
     })
     return results.rows[0]
   }
 
-  async getStatementsForReviewer(reportId: number) {
+  async getStatementsForReviewer(reportId: number): Promise<ReviewerStatement[]> {
```
Comment (andrewrlee) [inline server/data/statementsClient.ts:92]: I've aligned the return types of getStatementForReviewer and getStatementsForReviewer to reuse same type definitions

File: server/routes/maintainingReports/reviewer.ts [hunk 3]
```diff
@@ -78,27 +78,22 @@ export default class ReviewerRoutes {
...
 
-    const data = { incidentId: reportId, reporterName, submittedDate, offenderDetail, statements, tab }
-
     const reportDataForPrint = await this.reportDetailBuilder.build(res.locals.user.username, report)
 
-    const submittedStatements = statements.filter(stmnt => stmnt.isSubmitted)
-
-    const statementsWithNarrative = await Promise.all(
-      submittedStatements.map(statement => this.reviewService.getStatement(statement.id))
-    )
```
Comment (andrewrlee) [inline server/routes/maintainingReports/reviewer.ts:89]: simplified all this so it's only calling the service layer once

File: server/services/drafts/draftReportService.test.ts [hunk 4]
```diff
@@ -65,7 +67,7 @@ describe('getUoFReasonState', () => {
   })
 
   test('it should return when present', async () => {
-    ;(isReportComplete as jest.Mock).mockReturnValue(false)
```
Comment (andrewrlee) [inline server/services/drafts/draftReportService.test.ts:68]: got an eslint error around here

File: server/services/reviewService.ts [hunk 5]
```diff
@@ -29,20 +30,23 @@ const toIncidentSummary = (namesByOffenderNumber: NamesByOffenderNumber) => (
...
-// Surely this is defined somewhere useful? In a base set of typescript types perhaps?
-interface Predicate<T> {
-  (x: T): boolean
-}
+type Predicate<T> = (t: T) => boolean
 
 const offenderNameFilter = (nameToMatch?: string): Predicate<IncidentSummary> => {
-  const regexp = new RegExp(`^.*${nameToMatch}.*$`, 'i')
+  const safeString = nameToMatch.replace(/[.*+?^${}()|[\]\\]/g, '\\$&')
+  const regexp = new RegExp(`^.*${safeString}.*$`, 'i')
```
Comment (andrewrlee) [inline server/services/reviewService.ts:37]: Noticed that we were just accepting raw user provided string as regex - which would blow up if invalid regex passed and makes the service susceptible to regex bombs (by authenticated users)

File: server/data/statementsClient.ts (hunk 2 above)
Comment (GurnankCheema) [reply server/data/statementsClient.ts:92]: cool

File: server/services/drafts/draftReportService.test.ts [hunk 6]
```diff
@@ -83,14 +85,14 @@ describe('getUoFReasonState', () => {
...
   test('it should handle when absent', async () => {
-    ;(isReportComplete as jest.Mock).mockReturnValue(false)
+    isReportCompleteMock.mockReturnValue(false)
     draftReportClient.get.mockResolvedValue({ form: {} })
     const result = await service.getUoFReasonState('user1', 1)
     expect(result).toStrictEqual({ isComplete: false, primaryReason: undefined, reasons: [] })
   })
 
   test('when form is complete', async () => {
-    ;(isReportComplete as jest.Mock).mockReturnValue(true)
```
Comment (GurnankCheema) [inline server/services/drafts/draftReportService.test.ts:93]: yes, I never did like those leading semi colons!

File: server/services/reviewService.ts (hunk 5 above)
Comment (GurnankCheema) [reply server/services/reviewService.ts:37]: cool

**Review by GurnankCheema — APPROVED** [review]
LGTM

**Pull request merged**